File: sasmodels/models/lamellar_x5_paracrystal_kr.c
/*Lamellar_ParaCrystal 
July 2019 RKH - converted his FISH Fortran code for the Kotlarchyk & Ritzau form factor of sheets and structure factor
for 1d paracrystals into c++ for sasview. Ref.  J.Appl.Cryst. 24(1991)753-758 with some "corrections" as per version from FISH .
SHOULD put paraCryst_kr and rlayff_kr into a library as they could also be used by other routines
12 June 2020 initial version of lamellar12345 type models, (released privately to two user groups).
Sept 2023 this x5 version has explicit amounts of 1 through 5 layers, plus a stack with an adjustable number of layers
BEWARE P(Q) equations divide by zero if sigma_t_by_t =0.0 (and may not then give results even when reset!)
This is very flexible, particularly for mixtures of unilamellar and multilamellar particles.
IF sigma_d_by_d = -1 ignore the S(Q), return only P(Q) for sheet, including (scale1 +scale2+ ...)(del rho)^2
IF sigma_t_by_t = -1 ignore the P(Q), return only AVERAGE S(Q) sum(scale_n*S(Q,n) )/sum(scale_n), or of course a 
single S(Q,n) if the other scale_n are set to zero. 

*/
double paraCryst_kr(double qval, double RM, double D, double GD);
double rlayff_kr(double qval, double RLBAR, double GL, double TT, double RSIG, double f1);

static double
Iq(double qval,
   double thickness,
   double sigma_t_by_t,
   double interface_t,
   double rsig_lorentz,
   double scale1,
   double scale2,
   double scale3,
   double scale4,
   double scale5,
   double scaleMadj,
   double Madj,
   double davg,
   double sigma_d_by_d,
   double sld,
   double solvent_sld)
{
      double f1, f2, Znq2, Znq3, Znq4, Znq5, ZnqM, sqsum, contr;
      const double fp_2 = 2.0;
      const double fp_3 = 3.0;
      const double fp_4 = 4.0;
      const double fp_5 = 5.0;
      sqsum = 0.0;
      
/*      //get the fractional part of Nlayers, to determine the "mixing" of N's
// have removed this averaging of two values of Nlayers, as the pattern changes 
// dramatically if e.g. go from 3 to 3.5 to 4 layers  by this method, 
// the original varies more smoothly, even if a non-integer number of layers is "unphysical"!
      int n1 = (int)(fp_Nlayers);            //truncate towards zero
      int n2 = n1 + 1;
      const double xn = (double)n2 - fp_Nlayers;      //fractional contribution of n1
      //
      //calculate the n1 contribution

      Znq = xn*paraCryst_kr(qval,(double)n1,davg,sigma_d_by_d);
      //
      //calculate the n2 contribution
      Znq += (1.0-xn)*paraCryst_kr(qval,(double)n2,davg,sigma_d_by_d);
*/      
      // all these IF's will I'm told will reduce the efficiency of gpu code
      if (sigma_d_by_d >= -0.999){
          // normal route
          if(scale2 > 0.0){sqsum += scale2*paraCryst_kr(qval,fp_2,davg,sigma_d_by_d);}
          if(scale3 > 0.0){sqsum += scale3*paraCryst_kr(qval,fp_3,davg,sigma_d_by_d);}
          if(scale4 > 0.0){sqsum += scale4*paraCryst_kr(qval,fp_4,davg,sigma_d_by_d);}
          if(scale5 > 0.0){sqsum += scale5*paraCryst_kr(qval,fp_5,davg,sigma_d_by_d);}
          if(scaleMadj > 0.0){sqsum += scaleMadj*paraCryst_kr(qval,Madj,davg,sigma_d_by_d);}
      } else {
          // return only scaled P(Q)
          sqsum = scale2 +scale3 +scale4 +scale5 +scaleMadj;
      }
      // Could likely speed up a little with a bespoke paraCryst_kr for all five at once as parts of the
      // calculation are repeated.
      // One day separate the S(Q) and P(Q) here into individual models
      // meanwhile it should be relatively simple to make other versions with shell/core/shell layers etc.
      // I believe that <F> and F^2> are being properly averaged for the polydisperse layer
      
      if (sigma_t_by_t < -0.999){
          // return only average S(Q) which should tend to 1.0 at high Q
           return sqsum/(scale2 +scale3 +scale4 +scale5 +scaleMadj);
      } else if( sigma_t_by_t > 1.0e-08){
      contr = sld - solvent_sld;
           f2 = 0.25*contr*contr*rlayff_kr(qval,thickness,sigma_t_by_t,interface_t,rsig_lorentz,f1);
      } else{ 
      //  catch divide by zero when sigma_t_by_t = 0.0
      // TODO: add proper P(Q) for monodisperse sheet
      f2 =0.0;
      }
      //
      return ( scale1 + sqsum )*f2*1.0e-04;
      
}

// functions for the lamellar paracrystal model, converted from my own FISH fortran code, July 2019, RKH
double 
paraCryst_kr(double QQ, double RM, double D, double GD) {
    // presume RMU is mu=1 = cos(angle between Q and axis of stack) in eqn (13)
    const double RMU = 1.0;
    const double QDM = QQ*D*RMU;
    const double AA = -square(GD*QDM)*0.5;
    //      F=EXPSPEC(-0.5*(GD*QDM)**2)
    //      F2=F*F
    //      FPM=F**RM
    //  take the powers of F inside the exponential
    const double F = exp(AA);
    const double F2 = exp(2.0*AA);
    const double FPM = exp(RM*AA);
    //
    // this hits severe problems when Q goes small and F goes up towards
    //  1.0 from below.  We may have to argue that the cos(M*QDM) oscillates
    // faster than cos(QDM) so averages to zero ?
    //
    //      PARCRYS = (1.0-F2)/(1.0-2.*F*COS(QDM) +F2) +(1.0/RM)*
    //     > ( -2.0*F*( (1.0+F2)*COS(QDM) -2.0*F -FPM*COS(QDM*(RM+1.)) +
    //     > 2.0*FPM*F*COS(RM*QDM) -FPM*F2*COS( (RM-1.0)*QDM ) ) )/
    //     > ( 1.0 - 2.0*F*COS(QDM) + F2)**2
    //
    // rearranged, 13/2/92, gives same results
    // tested 14/2/92 in double precision, single prec. gave same results except
    // small errors at very low Q ( <.002), and through peak (<.00002),
    // with RM=10, D=100, GD=0.15 as typical parameters.
    //
    double Snq = (1.0/RM)*(4.0*F2 + RM*(1. - F2*F2) - 4.0*F2*FPM*cos(QDM*RM) + 
                  cos(QDM)*2.*F*(-RM*(1. - F2) - (1. + F2) + FPM*(1. + F2)*cos(RM*QDM)) +
                  sin(QDM)*2.*F*FPM*(F2 - 1.)*sin(RM*QDM))/square(1. - 2.0*F*cos(QDM) + F2);
    // for very low GD, and always at low Q the results tend to the 
    // equation for a perfectly ordered system:
    //      PARCRYS=(1.0/RM)*(SIN(0.5*RM*QDM)/SIN(0.5*QDM))**2
    return Snq;
}

double
rlayff_kr(double QQ, double RLBAR, double GL, double TT, double RSIG, double f1) {
    //// RKH July 2019, FORM FACTOR FOR POLYDISPERSE SHEET - Kotlarchyk & Ritzau, 
    ////  J.Appl.Cryst. 24(1991)753-758,  eqns (16) & (17) with 1/Q**2
    // included as per eqn (13)          12/2/92 RKH
    // interfacial thickness TT = (2*pi)**0.5 * SIG    CORRECTED 7/4/92
    // GL = SIGMA(L)/LBAR = (Z+1)**-0.5
    // 16/3/93 add extra parameter RSIG to model, as Lorentz factor,
    // see Skipper et.al. J.Chem.Phys 94(1991)5751-5760
    // Sept 2023, divide by RLBAR.RSIG^2 to try to get volume fraction phi scaling correct
    // should also reduce strong parameter correlation between these and scale parameters.
        double ZP1 = 1.0/square(GL);
        const double Z = ZP1 - 1.0;
        double ZM1 = Z - 1.0;
        const double AL = ZP1/(QQ*RLBAR);
        const double AL2 = 2.0*AL;
        const double DEN = (1.0 + square(QQ*RSIG)*0.5);
        const double Q2S2 = -square(QQ*TT)/(2.0*M_PI);
        //
        //     F=(AL2**ZP1)*SIN(Z*atan(1.0/AL2))*EXP(0.5*Q2S2)/
        //    >  (Z* ( ( AL2**2 +1)**(0.5*Z) )*QQ )
        //rearrange in attempt to avoid overflows ( note AL can be 100, Z say 25 )
        
        // sasview 4 does not need <F> but it is here ready .
        // I believe that <F> and F^2> are being properly averaged for the polydisperse layer
        double AL2SQ = square(AL2);
        // f1 = (1. + 1.0/AL2SQ)**(-0.5*ZP1)*sqrt(AL2SQ + 1)*sin(Z*atan(1.0/AL2))*exp(0.5*Q2S2)/Z;
        //
        f1 = 0.0;
        // start of final code for f1, comment out here as we are not using it anywhere
        //f1 = pow( 1.0 + 1.0/AL2SQ, -0.5*ZP1);
        //f1 *= sqrt(AL2SQ + 1)*sin(Z*atan(1.0/AL2))*exp(0.5*Q2S2)/Z;
        //f1 /= sqrt(DEN);
        // end of code for f1
        
        // original fortran
        //     RLAYFF= (AL2**ZP1)*( (AL2**(1.0-Z)) -((AL2**2 +4.0)**(-0.5*ZM1))*
        //    >   COS( ZM1*atan(1.0/AL) ) )*EXP(Q2S2)/ (2.0*Z*ZM1*QQ*QQ)
        
        //rearrange to avoid overflows ( note AL can be 100, Z say 25 )
        //included an extra factor of 4.0 to get beta(Q) to go to 1.0 at low Q and P(Q) to go to 1.0/Q**2
        // f2 = 4.0*(AL2SQ)*(1. - ((1. + 4.0/(AL2SQ))**(-0.5*ZM1))*cos(ZM1*atan(1.0/AL)))*exp(Q2S2)/(2.0*Z*ZM1);
        
        double f2 = 4.0*(AL2SQ)*(1. - pow(1. + 4.0/AL2SQ,-0.5*ZM1)*cos(ZM1*atan(1.0/AL))) * exp(Q2S2)/(2.0*Z*ZM1);
        f2 /= DEN;
        return 1.0e-04*(1.0 + square(RSIG))*RLBAR*f2;
}

File: sasmodels/models/lamellar_x5_paracrystal_kr.py
# Note: model title and parameter table are inserted automatically
#2345 789012345678901234567890123456789012345678901234567890123456789012
r"""

Implementation of Kotlarchyk & Ritzau paracrystal stack model, 
J.Appl.Cryst. 24(1991)753-758, which has Shultz polydisperse sheets 
with diffuse interfaces, multiplied by a paracrystalline struture 
factor. Additionally here a Lorentz term is included at small Q.to 
allow for some curvature or waviness of the layers.

This x5 version has explicit amounts of N = 1,2,3,4 or 5 layers, 
plus an additional stack with an adjustable number of M\_adj layers.

Thus a distribution in N may be designed and/or the number of layers 
fit. Beware that the best fit may be poorly determined, so do not 
adjust too many parameters at once! In practice this model has proven 
to be very effective in systems with a mixture of 
unilammellar and multilammellar vesicles.

The vesicles must be sufficiently large (or flexible) that the 
interference in Q space from opposite faces, across their diameter, is 
well below the minimum Q values of the data. If in doubt about this 
and/or to help understand both the absolute intensity and 
rsig\_lorentz distance, compare results from the :ref:`vesicle` model 
which uses the exact form factor for rigid spheres.

This is a first attempt at the equations, reports of any issues 
with them or the calculations in the code will be gratefully received.

Further explanations of the scale parameters will follow. It will 
be best to leave the overall scale parameter included by sasview set 
to 1.0 or some other convenient value.

Definition
----------

In the equations below,

- *scale* for each of $Ni = 1, 2, 3, 4, 5$ and then $Madj$ layer stacks determines 
   the amount of each in the final intensity,

- *sld* $-$ *sld_solvent* is the contrast $\Delta \rho$,

- *thickness* is the mean layer thickness, the same for all 
  layers $\langle t \rangle$,

- *sigma_t* is the standard deviation of the Schultz
  layer distance distribution from $\sigma_t / \langle t \rangle$.

- *interface_t* is the width of the interfaces, note $\sigma_{interface} = interface\_t/\sqrt{2\pi}$,

- *The number of layers* in each stack is $Ni = 1, 2, 3, 4, 5$, then $Madj >= 2$,

- *d_spacing* is the average distance between adjacent layers
  $d = \langle D \rangle$, its initial value will need to be close to $2\pi/Qpeak$ 
  for a fit to work,

- *sigma_d* is the standard deviation of the Gaussian
  layer distance distribution from $\sigma_D / \langle D \rangle$, and

- *rsig\_lorentz* in the Lorentz term allows approximately for a finite
  radius of curvature and/or lateral waviness in the layers, as its
  value increases from zero the scattering moves rapidly from the 
  $Q^{-2}$ dependence of scatter from an oriented flat sheet to the 
  $Q^{-4}$ dependence from randomly oriented sheets.


The scattering intensity $I(q)$ is calculated as

.. math::

    I(q) = \frac{\Delta\rho^2P_{layer}(q)}{4(1+rsig{\_}lorentz^2q^2/2)}( scale\_N1 + \sum_{i=2}^{5} \left[ scale\_Ni.S(q,i)\right] )


This will then be multiplied by the leading sasview scale parameter and a 
flat background added. The form factor of the layer, $P\_{layer}(q)$ follows 
eqn (17) in the reference (but has been re-arranged in the c++ code to help 
avoid computational errors). 

There is some uncertainty regarding the leading factor of 4, which now seems 
to cancel with the 4 in the numerator of the previous equation, see notes 
in c++ code.

.. math::

    P\_{layer}(q) = \frac{4(2\alpha^{z+1}(2\alpha^{(1-z)} -
       ((2\alpha)^2+4)^{-(z-1)/2})}{2z(z-1)}
       \cos((z-1)tan^{-1}(1/\alpha))\exp{(-q^2\sigma_{interface}^2/2)}

where

.. math::

    z+1 = \frac{1}{(\sigma_t/t)^2}

and

.. math::

    \alpha = \frac{(z+1)}{Qt} = \frac{1}{Qt(\sigma_t/t)^2}

The structure factor for each stack of $N$ layers follows equations (10-12) of 
the reference. (Symbols $I$ and $F$ here follow those in the reference, which are 
different to their more usual meanings.)

.. math::

    S(q,N) = Z(Q) + \frac{I_c(q,N)}{N}

where

.. math::

    I_c(q,N) = -2F&\biggl\{(1+F^2)\cos(qd)-2F-F^N\cos((N+1)qd) \\
    & +2F^{(N+1)}\cos(Nqd)-F^{(N+2)}\cos((N-1)qd) \biggr\}/(1-2F\cos(qd)+F^2)^2


where

.. math::

    Z(q) = \frac{(1-F^2)}{(1-2F\cos(qd)+F^2)}

and

.. math::

    F=exp{(-\sigma_d^2q^2/2)}


There is currently no 2d intensity calculation for this model.


Source
------

Reference
---------
"Paracrystal Model of the High-Temperature Lamellar Phase of a Ternary Microemulsion
System", M.Kotlarchyk & S.M.Ritzau, J.Appl.Cryst. 24(1991)753-758

Authorship and Verification
----------------------------

* **Author:** Richard Heenan **Date:** April 24, 2019
* **Last Modified by:Richard Heenan, Sept 2023**
* **Last Reviewed by:**
"""

import numpy as np
from numpy import inf

name = "lamellar_x5_paracrystal_kr"
title = "Random lamellar sheet plus stacks with paracrystal structure factors"
description = """\
    [lamellar sheet plus stacks with paracrystal structure factors]

"""
category = "shape:lamellae"
have_Fq = False
single = False

#             ["name", "units", default, [lower, upper], "type","description"],
parameters = [["thickness", "Ang", 33.0, [0, inf], "volume",
               "mean sheet thickness"],
              ["sigma_t_by_t", "", 0.1, [0.0, inf], "",
               "Sigma (relative polydispersity) of the sheet thickness"],
              ["interface_t", "Ang", 5.0, [0.0, inf], "",
               "interfacial thickness, sqrt(2.pi).sigma, damps form factor"],
              ["rsig_lorentz", "Ang", 100.0, [0.0, inf], "",
               "Lorentz factor for curvature or orientation"],
              ["scale_N1", "", 2.5, [0.0, inf], "",
               "relative amount of N = 1, monolayer"],
              ["scale_N2", "", 5.0, [0.0, inf], "",
               "relative amount of N = 2, bilayer"],
              ["scale_N3", "", 0.5, [0.0, inf], "",
               "relative amount of N = 3, three layers"],
              ["scale_N4", "", 0.0, [0.0, inf], "",
               "relative amount of N = 4, four layers"],
              ["scale_N5", "", 0.0, [0.0, inf], "",
               "relative amount of N = 5, five layers"],
              ["scale_Madj", "", 0.0, [0.0, inf], "",
               "relative amount of Madj layers"],
              ["Madj", "", 2.0, [2.0, 50], "",
               "adjustable number of layers"],
              ["d_spacing", "Ang", 75., [0.0, inf], "",
               "lamellar spacing of paracrystal stack"],
              ["sigma_d_by_d", "", 0.05, [0.0, inf], "",
               "Sigma (relative polydispersity) of the lamellar spacing"],
              ["sld", "1e-6/Ang^2", 1.0, [-inf, inf], "sld",
               "layer scattering length density"],
              ["sld_solvent", "1e-6/Ang^2", 6.34, [-inf, inf], "sld",
               "Solvent scattering length density"]
             ]

source = ["lamellar_x5_paracrystal_kr.c"]

form_volume = """
    return 1.0;
"""
#TODO test this random function
def random():
    scale_N1 = np.random.randint(0, 1)
    scale_N2 = np.random.randint(0, 1)
    scale_N3 = np.random.randint(0, 1)
    scale_N4 = np.random.randint(0, 1)
    scale_N5 = np.random.randint(0, 1)
    scale_Madj = np.random.randint(0, 1)
    Madj = np.random.randint(6, 50)
    thickness = np.random.uniform(10, 1000)
    d_spacing = thickness*np.random.uniform(1.2,5.0)
    # Let polydispersity peak around 15%; 95% < 0.4; max=100%
    sigma_d_by_d = np.random.beta(1.5, 7)
    pars = dict(
        thickness=thickness,
        scale_N1=scale_N1,
        scale_N2=scale_N2,
        scale_N3=scale_N3,
        scale_N4=scale_N4,
        scale_N5=scale_N5,
        scale_Madj=scale_Madj,
        Madj=Madj,
        d_spacing=d_spacing,
        sigma_d=sigma_d_by_d,
    )
    return pars

# ER defaults to 0.0
# VR defaults to 1.0

#TODO check the demo plot
demo = dict(scale=1, background=0,
            thickness=33, scale_N1=0.5, scale_N2=0.0, scale_N3=0.5, scale_N4=0.0, scale_Madj=0.75, Madj=40,
            d_spacing=250, sigma_d_by_d=0.2,
            sld=1.0, sld_solvent=6.34)

#TODO needs new unit tests here  
'''
# this from a different model - edit it!
tests = [
    [{'scale': 1.0, 'background': 0.0, 'thickness': 33., 'Nlayers': 20.0,
      'd_spacing': 250., 'sigma_d': 0.2, 'sld': 1.0,
      'sld_solvent': 6.34, 'thickness_pd': 0.0, 'thickness_pd_n': 40},
     [0.001, 0.215268], [21829.3, 0.00487686]],
]
'''


